crates/uplc/src/machine.rs

```diff
@@ -0,0 +1,407 @@
+    env: Vec<Value>,
+    unbudgeted_steps: [u32; 8],
+    pub logs: Vec<String>,
+}
+
+impl Machine {
+    pub fn new(costs: CostModel, initial_budget: ExBudget, slippage: u32) -> Machine {
+        Machine {
+            costs,
+            ex_budget: initial_budget,
+            slippage,
+            frames: vec![Context::NoFrame],
+            env: vec![],
+            unbudgeted_steps: [0; 8],
+            logs: vec![],
+        }
+    }
+
+    pub fn run(&mut self, term: &Term<NamedDeBruijn>) -> Result<Term<NamedDeBruijn>, Error> {
+        let startup_budget = self.costs.machine_costs.get(StepKind::StartUp);
+
+        self.spend_budget(startup_budget)?;
+
+        self.compute(term)
+    }
+
+    fn compute(&mut self, term: &Term<NamedDeBruijn>) -> Result<Term<NamedDeBruijn>, Error> {
+        match term {
+            Term::Var(name) => {
+                self.step_and_maybe_spend(StepKind::Var)?;
+
+                let val = self.lookup_var(name.clone())?;
+
+                self.return_compute(val)
+            }
+            Term::Delay(body) => {
+                self.step_and_maybe_spend(StepKind::Delay)?;
+
+                self.return_compute(Value::Delay(*body.clone()))
+            }
+            Term::Lambda {
+                parameter_name,
+                body,
+            } => {
+                self.step_and_maybe_spend(StepKind::Lambda)?;
+
+                self.return_compute(Value::Lambda {
+                    parameter_name: parameter_name.clone(),
+                    body: *body.clone(),
+                })
+            }
+            Term::Apply { function, argument } => {
+                self.step_and_maybe_spend(StepKind::Apply)?;
+
+                self.push_frame(Context::FrameApplyArg(*argument.clone()));
+
+                self.compute(function)
+            }
+            Term::Constant(x) => {
+                self.step_and_maybe_spend(StepKind::Constant)?;
+
+                self.return_compute(Value::Con(x.clone()))
+            }
+            Term::Force(body) => {
+                self.step_and_maybe_spend(StepKind::Force)?;
+
+                self.push_frame(Context::FrameForce);
+
+                self.compute(body)
+            }
+            Term::Error => Err(Error::EvaluationFailure),
+            Term::Builtin(fun) => {
+                self.step_and_maybe_spend(StepKind::Builtin)?;
+
+                let runtime: BuiltinRuntime = (*fun).into();
+
+                self.return_compute(Value::Builtin {
+                    fun: *fun,
+                    term: term.clone(),
+                    runtime,
+                })
+            }
+        }
+    }
+
+    fn return_compute(&mut self, value: Value) -> Result<Term<NamedDeBruijn>, Error> {
+        // avoid unwrap if possible and just return an err when None
+        // but honestly it should never be empty anyways because Machine
+        // is initialized with `Context::NoFrame`.
+        let frame = self.frames.last().cloned().unwrap();
```

Comment:
I'm not sure I follow the comment here. Are you intending to remove the `unwrap`?

Reply:
I need to update the comment, we'll probably just keep the unwrap. There will always at least be Context::NoFrame in the Vec

Reply:
Okay. Maybe use `expect()` instead of `unwrap()` to give context

Reply:
Thanks will do